### ISSUE TITLE
🔨 Implement command options (subcommands)

### ISSUE THREAD

**Issue by Arthurdw** [Pincer-org/Pincer#96]
## Feature

Implement subcommands. So that eg the following data gets sent to discord:

```json5
{
    "name": "permissions",
    "description": "Get or edit permissions for a user or a role",
    "options": [
        {
            "name": "user",
            "description": "Get or edit permissions for a user",
            "type": 2, // 2 is type SUB_COMMAND_GROUP
            "options": [
                {
                    "name": "get",
                    "description": "Get permissions for a user",
                    "type": 1, // 1 is type SUB_COMMAND
                    "options": [
                        {
                            "name": "user",
                            "description": "The user to get",
                            "type": 6, // 6 is type USER
                            "required": true
                        },
                        {
                            "name": "channel",
                            "description": "The channel permissions to get. If omitted, the guild permissions will be returned",
                            "type": 7, // 7 is type CHANNEL
                            "required": false
                        }
                    ]
                },
                {
                    "name": "edit",
                    "description": "Edit permissions for a user",
                    "type": 1,
                    "options": [
                        {
                            "name": "user",
                            "description": "The user to edit",
                            "type": 6,
                            "required": true
                        },
                        {
                            "name": "channel",
                            "description": "The channel permissions to edit. If omitted, the guild permissions will be edited",
                            "type": 7,
                            "required": false
                        }
                    ]
                }
            ]
        },
        {
            "name": "role",
            "description": "Get or edit permissions for a role",
            "type": 2,
            "options": [
                {
                    "name": "get",
                    "description": "Get permissions for a role",
                    "type": 1,
                    "options": [
                        {
                            "name": "role",
                            "description": "The role to get",
                            "type": 8, // 8 is type ROLE
                            "required": true
                        },
                        {
                            "name": "channel",
                            "description": "The channel permissions to get. If omitted, the guild permissions will be returned",
                            "type": 7,
                            "required": false
                        }
                    ]
                },
                {
                    "name": "edit",
                    "description": "Edit permissions for a role",
                    "type": 1,
                    "options": [
                        {
                            "name": "role",
                            "description": "The role to edit",
                            "type": 8,
                            "required": true
                        },
                        {
                            "name": "channel",
                            "description": "The channel permissions to edit. If omitted, the guild permissions will be edited",
                            "type": 7,
                            "required": false
                        }
                    ]
                }
            ]
        }
    ]
}
```

## Code implementation

Open for discussion, please post suggestions on how you'd like to use this in the lib.

## Useful link(s)

[https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-structure](https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-structure)

**Comment by Arthurdw** [comment]
@mwath yeah, like I said I'd like to do it through type annotations. _(So your `Using the function as type hint`)_. Which is a system that could work. But I also quote: `I just haven't figured out yet how to do groups properly.`

Thanks for the suggestions tho, and its great to see that we're on the same page. 😊

**Comment by Lunarmagpie** [comment]
Any ideas? I'd like to pick something so someone can get started on this cause i feel like this will never happen otherwise.
Right now im leaning torwards @mwath's third idea.

**Comment by Lunarmagpie** [comment]
newest idea
```python
cmd = Command(name='cmd', description=...)

@command(description="description", parent=cmd)
async def sub1(self, arg1: str):
    return arg1


@command(description="description", parent=cmd)
async def sub2(self, arg1: str, arg2: str):
    return arg1 + arg2
```
Using a parent argument will lead to less code duplication

**Comment by Lunarmagpie** [comment]
Different stolen implementations
[Hikari Lightbulb](https://hikari-lightbulb.readthedocs.io/en/latest/guides/command-groups.html?highlight=group#command-group-support)
tldr SUPER jank because the group is the second string in `@lightbulb.command` I rate 2/10
```python
import lightbulb

bot = lightbulb.BotApp(..., prefix="!")

@bot.command
@lightbulb.command("foo", "test group")
@lightbulb.implements(lightbulb.PrefixCommandGroup)
async def foo(ctx: lightbulb.Context) -> None:
    await ctx.respond("invoked foo")
```

[Hikari Tanjun](https://github.com/FasterSpeeding/Tanjun) not very bloated. I like this method.
```python
top_group = component.with_slash_command(tanjun.slash_command_group("places", "get info about places"))
nested_group = top_group.with_command(tanjun.slash_command_group("interaction", "say hello or something!"))

@nested_group.with_command
#A LOT MORE BLOAT HERE
```

**Comment by Lunarmagpie** [comment]
Could be cool
```python
class Bot(Client):
    @command(...)
    @group("")
    @subgroup("")
    def comamnd_here():
        ...
````

**Comment by Sigmanificient** [comment]
I still prefer the one from 18 days ago currently, but ig we can make a poll vote or something

**Comment by Lunarmagpie** [comment]
Guess ill go with that method since it seems really maintainable
